lib/gc/Transforms/GPU/GpuToGpuOcl.cpp

```diff
@@ -0,0 +1,560 @@
+
+namespace {
+LLVM::CallOp funcCall(OpBuilder &builder, const StringRef name,
+                      const Type returnType, const ArrayRef<Type> argTypes,
+                      const Location loc, const ArrayRef<Value> arguments,
+                      bool isVarArg = false) {
+  auto module = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  auto function = module.lookupSymbol<LLVM::LLVMFuncOp>(name);
+  if (!function) {
+    auto type = LLVM::LLVMFunctionType::get(returnType, argTypes, isVarArg);
+    function = OpBuilder::atBlockEnd(module.getBody())
+                   .create<LLVM::LLVMFuncOp>(loc, name, type);
+  }
+  return builder.create<LLVM::CallOp>(loc, function, arguments);
+}
+
+// Assuming that the pointer to the context is passed as the last argument
+// of the current function of type memref<anyType> with zero dims. When lowering
+// to LLVM, the memref arg is replaced with 3 args of types ptr, ptr, i64.
+// Returning the first one.
+Value getCtxPtr(const OpBuilder &rewriter) {
+  auto func =
+      rewriter.getBlock()->getParent()->getParentOfType<LLVM::LLVMFuncOp>();
+  return func.getArgument(func.getNumArguments() - 3);
+}
+
+struct Helper final {
+  LLVMTypeConverter &converter;
+  Type voidType;
+  Type ptrType;
+  Type idxType;
+  mutable std::unordered_set<std::string> kernelNames;
+
+  explicit Helper(MLIRContext *ctx, LLVMTypeConverter &converter)
+      : converter(converter), voidType(LLVM::LLVMVoidType::get(ctx)),
+        ptrType(LLVM::LLVMPointerType::get(ctx)),
+        idxType(IntegerType::get(ctx, converter.getPointerBitwidth())) {}
+
+  Value idxConstant(OpBuilder &rewriter, const Location loc,
+                    size_t value) const {
+    return rewriter.create<LLVM::ConstantOp>(
+        loc, idxType,
+        rewriter.getIntegerAttr(idxType, static_cast<int64_t>(value)));
+  }
+
+  void destroyKernels(OpBuilder &rewriter, Location loc,
+                      ArrayRef<Value> kernelPtrs) const {
+    auto size = idxConstant(rewriter, loc, kernelPtrs.size());
+    auto kernelPtrsArray =
+        rewriter.create<LLVM::AllocaOp>(loc, ptrType, ptrType, size);
+    for (size_t i = 0, n = kernelPtrs.size(); i < n; i++) {
+      auto elementPtr =
+          rewriter.create<LLVM::GEPOp>(loc, ptrType, ptrType, kernelPtrsArray,
+                                       idxConstant(rewriter, loc, i));
+      rewriter.create<LLVM::StoreOp>(loc, kernelPtrs[i], elementPtr);
+    }
+
+    funcCall(rewriter, GPU_OCL_KERNEL_DESTROY, voidType, {idxType, ptrType},
+             loc, {size, kernelPtrsArray});
+  }
+};
+
+template <typename SourceOp>
+struct ConvertOpPattern : ConvertOpToLLVMPattern<SourceOp> {
+  const Helper &helper;
+
+  explicit ConvertOpPattern(const Helper &helper)
+      : ConvertOpToLLVMPattern<SourceOp>(helper.converter), helper(helper) {}
+};
+
+struct ConvertAlloc final : ConvertOpPattern<gpu::AllocOp> {
+  explicit ConvertAlloc(const Helper &helper) : ConvertOpPattern(helper) {}
+
```

Comment:
How is this different from `ConvertAllocOpToGpuRuntimeCallPattern` in handling the memrefs and shapes? I'd expect the lowering to look just the same except for the runtime call signature/params.

Reply:
ConvertAllocOpToGpuRuntimeCallPattern didn't work for me. Check this test:

```LLVM
module attributes {gpu.container_module} {
  func.func @main() attributes {llvm.emit_c_interface} {
    %0 = "gpux.create_stream"() : () -> !gpux.StreamType
    %memref = "gpux.alloc"(%0) {operandSegmentSizes = array<i32: 0, 1, 0, 0>} : (!gpux.StreamType) -> memref<8xf32>
    return
  }
}
```

`gc-opt --convert-gpux-to-llvm` gives the following result:
```LLVM
module attributes {gpu.container_module} {
  func.func @main() attributes {llvm.emit_c_interface} {
    %0 = llvm.mlir.zero : !llvm.ptr
    %1 = llvm.mlir.zero : !llvm.ptr
    %2 = llvm.call @gpuCreateStream(%0, %1) : (!llvm.ptr, !llvm.ptr) -> !llvm.ptr
    %3 = llvm.mlir.constant(8 : index) : i64
    %4 = llvm.mlir.constant(1 : index) : i64
    %5 = llvm.mlir.zero : !llvm.ptr
    %6 = llvm.getelementptr %5[%3] : (!llvm.ptr, i64) -> !llvm.ptr, f32
    %7 = llvm.ptrtoint %6 : !llvm.ptr to i64
    %8 = llvm.mlir.constant(64 : i64) : i64
    %9 = llvm.mlir.constant(0 : i32) : i32
    %10 = llvm.call @gpuMemAlloc(%2, %7, %8, %9) : (!llvm.ptr, i64, i64, i32) -> !llvm.ptr
    %11 = llvm.mlir.undef : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)>
    %12 = llvm.bitcast %10 : !llvm.ptr to !llvm.ptr
    %13 = llvm.insertvalue %12, %11[0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %14 = llvm.bitcast %10 : !llvm.ptr to !llvm.ptr
    %15 = llvm.insertvalue %14, %13[1] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %16 = llvm.mlir.constant(0 : i64) : i64
    %17 = llvm.insertvalue %16, %15[2] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %18 = llvm.insertvalue %3, %17[3, 0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %19 = llvm.insertvalue %4, %18[4, 0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    return
  }
  llvm.func @gpuCreateStream(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  llvm.func @gpuMemAlloc(!llvm.ptr, i64, i64, i32) -> !llvm.ptr
}
```
Pay attention to these lines:
```LLVM
    %5 = llvm.mlir.zero : !llvm.ptr
    %6 = llvm.getelementptr %5[%3] : (!llvm.ptr, i64) -> !llvm.ptr, f32
```

Reply:
Now it's almost the same, except the static size calculation.